src/core/reducers/addons.js

```diff
@@ -286,10 +286,55 @@ export function createInternalAddon(
   return addon;
 }
 
-const initialState = {};
+type AddonID = string;
 
-export type AddonState = {
-  [addonSlug: string]: AddonType,
+export type AddonState = {|
+  addons: { [addonId: AddonID]: AddonType },
+  byGUID: { [addonGUID: string]: AddonID },
+  bySlug: { [addonSlug: string]: AddonID },
+|};
+
+const initialState: AddonState = {
+  addons: {},
+  byGUID: {},
+  bySlug: {},
+};
+
+export const getAddonByID = (
+  state: { addons: AddonState }, id: string
+): AddonType | null => {
+  if (!id) {
+    return null;
+  }
+
+  return state.addons.addons[id] || null;
+};
+
+export const getAddonBySlug = (
+  state: { addons: AddonState }, slug: string
+): AddonType | null => {
+  const addonId = state.addons.bySlug[slug];
+
+  return getAddonByID(state, addonId);
+};
+
+export const getAddonByGUID = (
+  state: { addons: AddonState }, guid: string
+): AddonType | null => {
+  const addonId = state.addons.byGUID[guid];
```

Comment:
This *might* need to use [`getGuid()`](https://github.com/willdurand/addons-frontend/blob/8b810ba3f9d18f729217144c38f6b6ae909538ce/src/core/reducers/addons.js#L121):

```js
const addonId = state.addons.byGUID[getGuid(addon)];
```

but I'm not sure. Does the Discopane work with your current patch? You would need to be using a build with `mozAddonManager` then install an add-on and refresh the page to test it because I think that's only when the bonkers `installations` reducer kicks in.

Reply:
This is how the Discopane API results come in: https://github.com/willdurand/addons-frontend/blob/8b810ba3f9d18f729217144c38f6b6ae909538ce/src/disco/api.js#L11

Reply:
Yes, it took me a while to figure out how the disco results were fetched and "normalized".

I think this selector should be as clean as possible and the `getuid` should be used to retrieve the `guid` that will later be passed to it.

Reply:
FYI I tested this patch with Disco Pane locally on a `mozAddonManager`-patched Nightly and it loaded add-ons and both themes and extensions could be installed and uninstalled, so it appears to work just fine.